java/arcs/core/data/expression/InferredType.kt

```diff
@@ -0,0 +1,123 @@
+
+import arcs.core.util.BigInt
+import kotlin.reflect.KClass
+
+/**
+ * Matches [widenAndApply] in Expression.kt
+ *
+ * The types are listed in terms of decreasing width, so that a type at position n can always
+ * be widened to a type at position n - 1.
+ */
+private val widenOrder = listOf<InferredType.Primitive>(
+  InferredType.Primitive.NumberType,
+  InferredType.Primitive.DoubleType,
+  InferredType.Primitive.FloatType,
+  InferredType.Primitive.BigIntType,
+  InferredType.Primitive.LongType,
+  InferredType.Primitive.IntType,
+  InferredType.Primitive.ShortType,
+  InferredType.Primitive.ByteType
+)
+
+/** Convert a [KClass] into an appropriate [InferredType] */
+fun KClass<*>.toType() = InferredType.Primitive.of(this)
+
+/** Widen a list of types to the largest type that encompasses all of them. */
+fun widen(vararg types: InferredType): InferredType.Primitive {
+  types.forEach {
+    require(it is InferredType.Numeric) {
+      "$it is not numeric"
+    }
+  }
+
+  for (type in widenOrder) {
+    if (types.contains(type)) {
+      return type
+    }
+  }
+
+  throw IllegalArgumentException("$types contains incompatible types.")
+}
+
+/**
+ * An [InferredType] is based on following a tree of DSL expressions from the leaves to the root,
+ * flowing types upward so that each node can be assigned a conservative bound on the types it
+ * may output.
+ */
+sealed class InferredType {
+  open fun isAssignableFrom(other: InferredType) = this == other
+  fun union(other: InferredType) = UnionType(setOf(this, other))
+
+  /** Represents all primitive types such as numbers, text, booleans, etc. */
+  sealed class Primitive(val kClass: KClass<*>) : InferredType() {
+    override fun isAssignableFrom(other: InferredType): Boolean = when {
+      this is Numeric && other is Numeric -> widen(this, other) == this
+      else -> this == other
+    }
+
+    object DoubleType : Primitive(Double::class), Numeric
+    object FloatType : Primitive(Float::class), Numeric
+    object BigIntType : Primitive(BigInt::class), Numeric
+    object LongType : Primitive(Long::class), Numeric
+    object IntType : Primitive(Int::class), Numeric
+    object ShortType : Primitive(Short::class), Numeric
+    object ByteType : Primitive(Byte::class), Numeric
+    object NumberType : Primitive(Number::class), Numeric
+    object BooleanType : Primitive(Boolean::class)
+    object TextType : Primitive(String::class)
+    object NullType : Primitive(NullType::class)
+
+    companion object {
+      private val kclassToType by lazy {
+        listOf(
+          DoubleType,
+          FloatType,
+          BigIntType,
+          LongType,
+          IntType,
+          ShortType,
+          ByteType,
+          NumberType,
+          BooleanType,
+          TextType,
+          NullType
+        ).associateBy({ it.kClass }, { it })
+      }
+
+      /** Given a [KClass] return the corresponding [InferredType]. */
+      fun of(kClass: KClass<*>) = kclassToType[kClass]!!
```

Comment:
`require` over `!!`, please.

Reply:
Is this here just to get the types to agree? If so, there might be another way to go about it.